Enhance -split operator with RightToLeft splitting

IISResetMe opened the pull request:
This pull request adds RightToLeft support to the binary -split operator, as requested in #4765 

I've added tests for negative integer input for the `Max-substrings` operand to work with both regex patterns and scriptblocks.

SteveL-MSFT committed:
wrapped STA code sections in `#if STAMODE` for potential future use
removed -importsystemmodules switch from powershell.exe and related code
removed -consolefile parameter from powershell.exe and related code

SteveL-MSFT committed:
removed font and codepage handling code that is only applicable to Windows PowerShell
replaced LengthInBufferCells() method with Unicode adapted code from PSReadline

SteveL-MSFT committed:
removed `#if CORECLR` statements merging code base

SteveL-MSFT committed:
[feature]
removed code to show a GUI prompt for credentials as PSCore6 prompts in console

IISResetMe committed:
Add tests for negative -split limit

This commit adds tests for negative limits to the -split operator. The intended change to allow -split to accept a negative integer as a limit for "backwards" (right-to-left) splitting should work both when splitting using a scriptblock/predicate or a regex pattern

IISResetMe committed:
Add tests for negative -split limit

This commit adds tests for negative limits to the -split operator. The intended change to allow -split to accept a negative integer as a limit for "backwards" (right-to-left) splitting should work both when splitting using a scriptblock/predicate or a regex pattern

IISResetMe committed:
Add right-to-left support to the -split operator

This commit adds the RightToLeft option to the Regex instance in SplitWithPattern() when the limit value specified by the user is negative. Using the RightToLeft options would ensure minimal overhead to existing cases with a positive or no limit.

SplitWithPredicate() is modified to inspect characters from right to left by introducing a cursor from which the strIndex derives its value. Works well but requires explicit checking and reversing of the string buffer before populating the resulting list when working right-to-left.

IISResetMe committed:
Add right-to-left support to the -split operator

This commit adds the RightToLeft option to the Regex instance in SplitWithPattern() when the limit value specified by the user is negative. Using the RightToLeft options would ensure minimal overhead to existing cases with a positive or no limit.

SplitWithPredicate() is modified to inspect characters from right to left by introducing a cursor from which the strIndex derives its value. Works well but requires explicit checking and reversing of the string buffer before populating the resulting list when working right-to-left.

iSazonov commented on test/powershell/Language/Operators/SplitOperator.Tests.ps1:
```diff
@@ -30,13 +30,18 @@ Describe "Split Operator" -Tags CI {
-            $res[2] | Should Be "c"
-            $res[3] | Should Be "d"
+            $res.count | Should Be 1
+            $res[0] | Should Be "a b c d"
```
I believe we want to get "d" as a last element.
/cc @mklement0

IISResetMe replied:
While that might seem intuitive, it would make the `Max-substrings` operand asymmetrical. 

The guiding principal behind the change is that the sign (`+/-`) signifies only the direction of the split operation, whereas the absolute value determines the maximum resulting number of substrings, with 0 preserving the meaning of "unlimited"

mklement0 replied:
Actually, I wasn't thinking of _direction_, but _what end_ to apply the split to - elements returned should always reflect the input order.

In this case, due to `-1`, the _last 1_ element should be returned, preceded by the remaining _prefix_, so we should get:

```none
$rest, $result = 'a b c d' -split '', -1
$rest | Should Be 'a b c'
$result.Count | Should be 1
$result[0] | Should Be 'd'
```

IISResetMe replied:
Whether you call it *direction* or not doesn't change the fact that by splitting from the *end* rather than the *start* we are in fact modifying the direction of the operation :-) 

Wouldn't your proposed result be confusing given the name of the operand (`Max-substrings` rather than `Number-Of-Times-To-Split` or similar)?

mklement0 replied:
_Direction_ is ambiguous, as that could be interpreted as wanting to also reverse the order in which the resulting elements are returned (e.g., `'a b c d' -split ' ', -2` resulting in `'a b', 'd, 'c'`, which we don't want), so I wanted to clarify.

To me, `Max-Substrings` makes sense: a negative value means: return `Max-Substrings` elements _from the end_ (in input order).

mklement0 replied:
Re `Max-Substrings`: I see your point now. With my proposal, you actually get `Max-SubStrings + 1` elements back - this is the asymmetry you mention.

mklement0 replied:
I agree that your approach is better for consistency with how `-split` currently works.

Given that what was approved was my approach, however, we should get approval again: cc @SteveL-MSFT

IISResetMe replied:
If you think about how negative array indices are treated in PowerShell, it makes perfect sense, and as https://github.com/PowerShell/PowerShell/pull/5125/commits/1d59be0f17e3b4e4b6c5d57daaf882b364f2939f shows, shifting to your original suggestion is trivial, but I would like to keep it symmetric (again, given that the second operand is not an index but a boundary value)

mklement0 replied:
Understood and agreed.

I indeed originally had array-index logic in mind, but neglected to consider that the current `-split` behavior uses the-number-of-elements-I-want-plus-one-for-the-rest logic.

Just to spell it out: with your approach (which we should go with), both `1` and `-1` are no-ops - both request no splitting at all, and which end to apply it to is then moot.

SteveL-MSFT replied:
I agree that the symmetric behavior makes sense and updating the documentation with good examples will clear up any confusion. cc @lzybkr

iSazonov replied:
@mklement0 Could you please modify your Issue description to satisfy the behavior?

mklement0 replied:
@iSazonov: Thanks for reminding me - done.

iSazonov commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -681,24 +696,30 @@ private static object SplitWithPredicate(ExecutionContext context, IScriptExtent
                 }
 
+                if(rightToLeft)
+                    split.Reverse();
```
Can we exclude `reverse`?

IISResetMe replied:
Remove the `List<string>.Reverse()` call for performance reasons or...? 

Removing it would funtionally result in the substrings for each input being returned in the wrong order, ie:

```
PS> 'a b c d' -split {$_ -eq ' '},-3
d
c
a b
```

Rather than the expected:

```
PS> 'a b c d' -split {$_ -eq ' '},-3
a b
c
d
```

I see two possible alternatives: 
1. Change all invocations of `split.Add()` to `split.Insert()` conditional on `rightToLeft`
2. Implement two completely separate routines for negative and positive `limit` values, duplicating the code slightly (but keeping performance overhead at a minimum)

iSazonov replied:
Yes, my thoughts was about performance. I don't know what case is better. I guess we would use only one routine. At a minimum, we have to start with this.

IISResetMe replied:
Splitting it into 2 loops would mean getting rid of at least one non-local variable assignment per input item and a number of branch points, the intent (splitting from the end) would also be clearer. Any thoughts @lzybkr ?

lzybkr replied:
`List.Reverse` is not expensive, it swaps in place, see the [implementation](http://source.dot.net/#System.Private.CoreLib/src/System/Array.cs,b242f3290263f0f2) (List is really an Array).

Insert would be more expensive, shifting everything on each insert instead of swapping once like in Reverse.

iSazonov replied:
We could do reverse copy in next line in `ExtendList`.
By the way, what are we doing this recopy?

iSazonov replied:
Up.

IISResetMe committed:
Change -split negative limit behavior to act like array indices

On request in #5125 behavior is changed to allow -n to signify n substrings from the end + remainder of string.

IISResetMe committed:
Change -split negative limit behavior to act like array indices

On request in #5125 behavior is changed to allow -n to signify n substrings from the end + remainder of string.

IISResetMe committed:
Revert "Change -split negative limit behavior to act like array indices"

This reverts commit 1d59be0f17e3b4e4b6c5d57daaf882b364f2939f.

IISResetMe committed:
Revert "Change -split negative limit behavior to act like array indices"

This reverts commit 1d59be0f17e3b4e4b6c5d57daaf882b364f2939f.

iSazonov commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -637,6 +637,12 @@ private static object SplitWithPredicate(ExecutionContext context, IScriptExtent
             {
                 List<string> split = new List<String>();
```
It seems we could move this to line 653 as
```c#
List<string> split = new List<String>(limit); 
```

IISResetMe replied:
I'd keep it without specifying capacity up front, because you'd still need to support unbounded capacity in the special case of (`limit == 0`) and the (very common) case of _no `Max-substrings` specified_.

PetSerAl replied:
@iSazonov What if someone put `1000000` as `limit`? I do not think you should pre-allocate so much memory.

iSazonov replied:
I skipped ` limit == 0 `
Closed.

powercode replied:
Why shouldn't we specify the limit in the ctor? 
I'd gladly have powershell allocate the list once to the size I had given than have the list being reallocated and copied many times when the value is larger?
If someone specifies `1000000` they will get a list of roughly 8MB, which would soon be garbage collected, but with a single allocation, instead of the 19 we would get with the 0 default capacity.

IISResetMe replied:
Fair point. We could also allocate the list with capacity `Min(limit, item.Length +2)` (unless there's some negative-width split pattern I can't think of)

iSazonov replied:
I agree - it is good. Reallocation is much worse.

IISResetMe replied:
Forget about that `Min(limit, item.Length +2)` comment above, I just realized we're talking about regex with capture support, including `'abcde' -split '((?=.)(.)(?<=.))'` similarly terrifying things.

I've added the suggested changes + comments for the `List.Reverse()` call in https://github.com/PowerShell/PowerShell/pull/5125/commits/c423d8d6c998dd01ca9e7e2c54752d635bd7dd78

lzybkr replied:
I still like your thought to use `Min(limit, item.Length +2)`. It covers the common case sufficiently and protects against complete stupidity.

IISResetMe replied:
Just had a look at the [`List<T>` source](https://referencesource.microsoft.com/#mscorlib/system/collections/generic/list.cs,51), the default capacity reallocation behavior (multiply by 2) actually fits our needs perfectly if we do `Min(limit, item.Length + 2)` - matches worst case with no captures, and each nested capture-group added will at most require 1 extra re-allocation.

iSazonov replied:
[CoreFX link on EnsureCapacity](https://github.com/dotnet/corefx/blob/bffef76f6af208e2042a2f27bc081ee908bb390b/src/System.Collections/src/System/Collections/Generic/List.cs#L433)

travisty- committed:
StopJobCommand: Remove Unreachable Code (#5091)

This code will never get executed, and its logic is
duplicated in the HandleStopJobCompleted() method.

markekraus committed:
Add Uservoice Link to Issue Template (#5100)

SteveL-MSFT committed:
Change location of ModuleAnalysisCache so it isn't shared with Windows PowerShell (#5133)

Currently, it's using the same location as Windows PowerShell and the two should not be sharing one as the PSModulePaths are different between the two.

SteveL-MSFT committed:
Remove unncessary method

iSazonov committed:
Merge pull request #4995 from SteveL-MSFT/remove-apartmentstate

* wrapped STA code sections in `#if STAMODE` for potential future use
removed -importsystemmodules switch from powershell.exe and related code
removed -consolefile parameter from powershell.exe and related code

* removed font and codepage handling code that is only applicable to Windows PowerShell
replaced LengthInBufferCells() method with Unicode adapted code from PSReadline

* removed `#if CORECLR` statements merging code base

* [feature]
removed code to show a GUI prompt for credentials as PSCore6 prompts in console

* Remove unncessary method 'LengthInBufferCellsFE'

JamesWTruher committed:
Make sure that there's a statement separator after we update the path, otherwise the built script will not run (#5137)

markekraus committed:
Add Get-WebListenerUrl Based Examples to WebListener README.md (#4981)

- Add Get-WebListenerUrl Based Examples to WebListener README.md
- Fix Cert test example

SteveL-MSFT committed:
Rename powershell.exe to pwsh.exe (#5101)

- Rename powershell.exe to pwsh.exe
- Fixe appveyor.psm1
- Update MSI to include 'pwsh' in path and app paths
- Revert change for hyper-v powershell direct
- Update names in packaging.psm1.
- Fix check for SxS

markekraus committed:
Add Authentication Parameter to Web Cmdlets for Basic and OAuth (#5052)

Closes #4274

Adds an -Authentication parameter to Invoke-RestMethod and Invoke-WebRequest
Adds an -Token parameter to Invoke-RestMethod and Invoke-WebRequest
Adds an -AllowUnencryptedAuthentication parameter to Invoke-RestMethod and Invoke-WebRequest
Adds tests for various -Authorization uses
-Authentication Parameter has 3 options: Basic, OAuth, and Bearer
Basic requires -Credential and provides RFC-7617 Basic Authorization credentials to the remote server
OAuth and Bearer require the -Token which is a SecureString containing the bearer token to send to the remote server
If any authentication is provided for any transport scheme other than HTTPS, the request will result in an error. A user may use the -AllowUnencryptedAuthentication switch to bypass this behavior and send their secrets unencrypted at their own risk.
-Authentication does not work with -UseDefaultCredentials and will result in an error.
The existing behavior with -Credential is left untouched. When not supplying -Authentication, A user will not receive an error when using -Credential over unencrypted connections.

Code design choice is meant to accommodate more Authentication types in the future.

Documentation Needed

The 3 new parameters will need to be added to the Invoke-RestMethod and Invoke-WebRequest documentation along with examples. Syntax will need to be updated.

dee-see committed:
Fix MatchInfoContext clone implementation (#5121)

The old implementation always set the new object's properties to null.

SteveL-MSFT committed:
removed RunspaceConfiguration support (#4942)

For PSCore 6, we are only supporting InitialSessionState. The RunspaceConfiguration APIs were already made internal. This PR removes all the code related to RunspaceConfiguration. This also means that some public APIs have changed. Was deciding between leaving the RunspaceConfiguration parameters and throwing Unsupported, but thought it was better to have it a compile-time error. This should simplify the code base.

IISResetMe committed:
Specify split List capacity when possible

This commit changes instantiatiation of the list used for substring chunks in SplitWithPredicate(), to have a specific capacity whenever known up front.

Fixed inconsistent bracing style with if/else as well

IISResetMe committed:
Specify split List capacity when possible

This commit changes instantiatiation of the list used for substring chunks in SplitWithPredicate(), to have a specific capacity whenever known up front.

Fixed inconsistent bracing style with if/else as well

chunqingchen committed:
Exclude PSHostProcess cmdlets from Unix platforms (#5105)

- Exclude 'Get-PSHostProcessInfo', 'Enter-PSHostProcess' and 'Exit-PSHostProcess' from Unix platforms.
- Update tests and add additional cmdlets need to be excluded

TravisEz13 committed:
Add compiler switches replace dangerous function with safer ones. (#5089)

* Add compiler switches replace the dangerous function with safer ones.
* Add -02 and make linker flags per platform
* add -O2 for arm and add -z,now
* remove windows linker flags

daxian-dbw committed:
Fetch resource string correctly (#5114)

thezim committed:
Add macos launcher. (#5138)

markekraus committed:
Replace HttpBin.org/response-headers Tests with WebListener (#5058)

SteveL-MSFT committed:
[feature] (#4960)

remove -computer support since corefx Process.GetProcesses(computer) returns local processes

rkeithhill committed:
Add document about how to create cmdlet w/dotnet CLI (#5117)

daxian-dbw committed:
Update appimage.sh to reflect the new name 'pwsh' (#5172)

SteveL-MSFT committed:
Enable import-module to be case insensitive (#5097)

Fix #1621 

Enable import-module to be case insensitive as macOS is case insensitive.

daxian-dbw committed:
Update the man help file used in packaging (#5173)

* Update pacakging.psm1 and rename man file

* Update pwsh.1.ronn

thezim committed:
Update powershell to pwsh. (#5174)

dlwyatt committed:
Unwrap 'ValueFromRemainingArguments' if the single element is a collection (#5109)

Unwrapping of 'ValueFromRemainingArguments' was being performed only for 'object[]' arrays (which covers the most common PowerShell binding scenarios) but could be skipped if a collection of any other type were passed to the parameter. This change unwraps 'ValueFromRemainingArguments' if the single element is a collection.

dantraMSFT committed:
Add exports for syslog APIs in 'libpsl-native'. (#5149)

TravisEz13 committed:
Add code to send WebHook for Travis-CI Daily Build (#5183)

markekraus committed:
Make -NoTypeInformation Default on Export-Csv and ConvertTo-Csv (#5164)

•Sets  -NoTypeInformation  as the default behavior for  Export-Csv  and  ConvertTo-Csv 
•Hides the  -NoTypeInformation  parameter switch
•Adds  -IncludeTypeInformation  switch to  Export-Csv  and  ConvertTo-Csv  to enable legacy behavior
•Provides a terminating error when both  -NoTypeInformation  and  -IncludeTypeInformation  are supplied
•adds tests for the new behavior
•fixes existing tests to align with new behavior

The new behavior will need to be documented.

daxian-dbw committed:
Update the help text with the new name 'pwsh' (#5182)

IISResetMe commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -681,24 +715,37 @@ private static object SplitWithPredicate(ExecutionContext context, IScriptExtent
+                    // original string (always output chunks left-to-right)
+                    split.Reverse();
+                }
                 ExtendList(results, split);
```
We _can't_ copy in reverse order without doing `split.Insert(0,item)` for each chunk. As @lzybkr noted earlier `List<T>.Insert()` is much more costly (it shifts the underlying array by copying every time) than simply adding in the order of the cursor and calling `Reverse()` before adding.

iSazonov replied:
I meant:
```c#
        private static void ExtendListReverse<T>(IList<T> list, IList<T> items)
        {
            for(int i=items.Count-1; i >= 0 ; i++ )
            {
                list.Add(items[i]);
            }
        }
```

powercode replied:
LGTM

IISResetMe replied:
@iSazonov Ahh, that makes sense (I completely misunderstood your request). While testing, I realized that I inadvertently broke the operation for subsequent strings when working on arrays due to the [modification of the `limit` param](https://github.com/PowerShell/PowerShell/pull/5125/files/5cf38f49940b3c31ad51088e2962186389c45939#diff-ebfc1e729c1362fdfc74174d6e5bafaeR643), resulting in:

```
PS C:\> 'a b c','d e f' -split {$_ -eq ' '},-2
a b
c
d
e f
```

Will update test cases and copy `Abs(limit)` to a local var

iSazonov replied:
Please add the new test.

IISResetMe replied:
Test cases added, `limit` bug fixed. 

I decided against implementing `ExtendListReverse()` for two reasons:
1. It's the only current use case
2. The name `ExtendListReverse()` seems quite ambiguous

iSazonov replied:
If we're starting to optimize, let's finish it.
We can use the name  `ReverseExtendList`.

I know it's a long time PR. Thank you for your patience. I think after my last comments, we're close to merge..

IISResetMe replied:
No worries, I want to get it right, this is a nice learning experience 👍 

Please see my latest comment re: `Reverse` of both the list and string buffer

IISResetMe commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -644,10 +650,31 @@ private static object SplitWithPredicate(ExecutionContext context, IScriptExtent
+                    split = new List<String>(limit);
+                }
+                
                 StringBuilder buf = new StringBuilder();
```
Required capacity for the final list of substrings is easy to predict with a reasonable degree of confidence based on length alone. 

Required capacity for the individual substrings is near-impossible to predict, as we know nothing about the character distribution up front.

iSazonov replied:
My thoughts was about `StringBuilder buf = new StringBuilder(item.Length); `.
Another thought is that we recreate StringBuilder in cycle. I think we can allocate it before the external cycle `foreach (string item in content)` and reuse in the line with buf.Clear();

IISResetMe committed:
Specify initial capacity in SplitWithPredicate()

In case no Max-substrings argument is passed when calling -split with a predicate, set the capacity of the resulting list to item.Length + 1 to avoid re-allocating the underlying array multiple times

IISResetMe committed:
Specify initial capacity in SplitWithPredicate()

In case no Max-substrings argument is passed when calling -split with a predicate, set the capacity of the resulting list to item.Length + 1 to avoid re-allocating the underlying array multiple times

IISResetMe commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -644,10 +650,31 @@ private static object SplitWithPredicate(ExecutionContext context, IScriptExtent
+                List<string> split; 
+                if(limit == 0)
+                {
+                    split = new List<String>();
```
Reviewed the code again, realized that the scenario I was trying to predict above doesn't actually apply to SplitWithPredicate(), worst case is `item.Length + 1`. I've added this to the constructor call when `limit == 0`.

iSazonov replied:
Closed.

IISResetMe committed:
Add test cases for Split with predicate against array

-split supports array input, but using -split with a scriptblock doesn't currently have any tests for array input

IISResetMe committed:
Add test cases for Split with predicate against array

-split supports array input, but using -split with a scriptblock doesn't currently have any tests for array input

IISResetMe committed:
Undo modification of limit parm in SplitWithPredicate()

A previously introduced modification to the limit parameter caused subsequent strings to be processed left-to-right when using negative Max-substrings against array input. Moved to local variable maxSubstrings

IISResetMe committed:
Undo modification of limit parm in SplitWithPredicate()

A previously introduced modification to the limit parameter caused subsequent strings to be processed left-to-right when using negative Max-substrings against array input. Moved to local variable maxSubstrings

IISResetMe commented:
travis build on linux seems to have failed for reasons unrelated to the code, how to re-run without commiting?

iSazonov commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -635,19 +635,49 @@ private static object SplitWithPredicate(ExecutionContext context, IScriptExtent
+                    split = new List<String>(maxSubstrings);
+                }
+                
                 StringBuilder buf = new StringBuilder();
```
We add a char by char in the `buf`. Doesn't make sense to recreate the `buf` after every separator.
Please move out of the cycle to the line 635 `StringBuilder buf = new StringBuilder(256); ` (default 16 is very short for typical strings) and replace the line 691 `buf = new StringBuilder();` on `buf.Clear();`.

IISResetMe replied:
This is fixed now. `StringBuilder buf` declaration moved outside loop, constructor calls replaced with `.Clear()`

iSazonov commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -681,24 +718,37 @@ private static object SplitWithPredicate(ExecutionContext context, IScriptExtent
-                        buf.Append(item[strIndex]);
+                        if(rightToLeft)
+                        {
+                            buf.Insert(0, item[strIndex]);
```
We add a char by char here so this is exactly the case when we should avoid `Insert`.
Please here remove `if(rightToLeft) ` and do only `buf.Append(item[strIndex]);` and add in the line 690
```
if(rightToLeft)
{
    buf.Reverse();
}
split.Add(buf.ToString());
buf.Clear();

IISResetMe replied:
I wanted to do that when we first dicussed the same issue with the `List<string>`, but unfortunately there's no such thing as `StringBuilder.Reverse()` :-\

iSazonov replied:
I think we can replace  `StringBuilder` with `List<Char>`.

IISResetMe committed:
Revert maxSubstrings and move r-to-l logic out of loop

Simplified logic for Right-to-left splitting by moving Abs(limit) assignment out of loop. Pre-emptive return on limit == 1 moved out as well

IISResetMe committed:
Revert maxSubstrings and move r-to-l logic out of loop

Simplified logic for Right-to-left splitting by moving Abs(limit) assignment out of loop. Pre-emptive return on limit == 1 moved out as well

IISResetMe committed:
Reuse existing StringBuilder in SplitWithPredicate()

Re-instantiating the string buffer in SplitWithPredicate() seems unnecessary, moved declaration outside loop, replaced reassignments with buf.Clear()

IISResetMe committed:
Reuse existing StringBuilder in SplitWithPredicate()

Re-instantiating the string buffer in SplitWithPredicate() seems unnecessary, moved declaration outside loop, replaced reassignments with buf.Clear()

IISResetMe commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -635,19 +635,49 @@ private static object SplitWithPredicate(ExecutionContext context, IScriptExtent
             foreach (string item in content)
             {
-                List<string> split = new List<String>();
+                int maxSubstrings = System.Math.Abs(limit);
```
Moved `rightToLeft` assignment outside loop, reverted to `limit = Abs(limit)`

IISResetMe committed:
Cleanup whitespace in SplitOperator.tests.ps1

IISResetMe committed:
Cleanup whitespace in SplitOperator.tests.ps1

IISResetMe commented:
`buf` re-allocation, `maxSubstrings`/`limit` and whitespace issues have all been addressed now. 

I still disagree on `Reverse()` + `ExtendList()` v `ReverseExtendList()` - won't add any more clarity to the code and doesn't seem justifiable with this single very specific use case.

---

I'm still trying to work out an elegant way of doing `buf.Reverse()` - thought about an extension method, like:

```
private static StringBuilder Reverse(this StringBuilder chunk)
{
    string str = chunk.ToString();
    if (str.Length < 2) 
        return chunk;
    char[] charArray = str.ToCharArray();
    Array.Reverse(charArray);

    return new StringBuilder(new string(charArray));
}
```

But I'm not sure that's the best way to go about it for multiple reasons:

1. Given I won't have access to the underlying array, I'll have to materialize the string before reversing it
2. StringBuilder ctor doesn't take `char[]`, so need to construct a new string again after reversing (lesser concern) 
3. Unlike existing methods (`Append*()`, `Insert()`, `Remove()` etc) the extension method won't actually change state of existing `chunk`, only return a new (reversed) `StringBuilder`

Any thoughts, recommendations?

powercode commented:
StringBuilder has an ```System.Text.StringBuilder Append(char[] value, int startIndex, int charCount)``` overload.
If you create the stringbuilder with value.Length as intialCapacity and use that append overload you get away from the allocation of the string

iSazonov commented:
I think in the context we can easy replace  StringBuilder  with  List<Char> .

IISResetMe commented:
Nice one @powercode, hadn't considered that. I think @iSazonov is right though. There's no inherent value in using `StringBuilder` over `List<char>` in the current implementation, and `List<char>` gives us a fairly efficient `Reverse()` option for free

IISResetMe committed:
Change buffer to List<char> in SplitWithPredicate

In order to avoid having to either call buf.Insert() or implement a custom Reverse() method for StringBuilder, I've replaced the char buffer in SplitWithPredicate() with a List<char>

IISResetMe committed:
Change buffer to List<char> in SplitWithPredicate

In order to avoid having to either call buf.Insert() or implement a custom Reverse() method for StringBuilder, I've replaced the char buffer in SplitWithPredicate() with a List<char>

IISResetMe commented:
`StringBuilder` replaced with `List<char>`, and I've removed the `Insert()` call in favor of `Reverse()` prior to adding the chunk to the list.

IISResetMe commented:
Any quick advice on how to do it correctly @TravisEz13 ? Last time I tried [it went horribly wrong](https://github.com/PowerShell/PowerShell/pull/5026#issuecomment-336746060)

IISResetMe committed:
Merge branch 'enhance-Split-RTL-1' of https://github.com/IISResetMe/PowerShell into enhance-Split-RTL-1

IISResetMe commented:
Alright, I rebased like you suggested, then did a pull and a push (as there were no merge conflicts), but it still dragged all the intermediate commits from the master branch into my PR :-\ @TravisEz13

iSazonov commented:
I use 
```
git chechout <branch>
git pull --rebase PowerShell master
```
I think you need remove the merge commit and try rebase again.

IISResetMe commented:
@TravisEz13 @iSazonov Tried to revert the last merge locally with 

    git revert c8eb3407ac839e5191b99c3a719a7629d9fdd746 -m 2

but this just adds a new commit reverting all the changes of the others at once, super messy. How can I "reset" the branch (`--reset hard [latest PR-speficic commit]` doesn't seem to work because of the previous rebase/fast-forwarding)? 

Maybe it's easier to just make a diff patch of the relevant files, create a new branch and submit a new PR?

Apologies for the git n00bness

SteveL-MSFT commented:
@IISResetMe the easier thing is probably to copy off the files you changed, reset this branch to before your first commit, copy over your changed files, then rebase off upstream/master, and push those changes.

If you want to do it the `git` way, you should `git rebase -i HEAD~56` (for your 56 commits), drop all the commits that aren't yours.  As it rebases, you'll hit lots of merge conflicts.  You'll have to fix them, `git add <file>` then `git rebase --continue` until it's clean.  Then `git fetch upstream` and `git rebase upstream/master` and resolve any merge conflicts.

iSazonov commented:
You can remove last commits by

git reset --hard ooooooo

IISResetMe commented:
Talked to @SteveL-MSFT offline, and decided to start over and push the changes to a new branch with the following commits:

https://github.com/IISResetMe/PowerShell/commit/81873b30f061056710a7417d7aafa4f091f6e1f9 (tests)
https://github.com/IISResetMe/PowerShell/commit/5e5e3e2f3c643cef73cb21968c58f4f3fc0f5a6c (SplitWithPattern())
https://github.com/IISResetMe/PowerShell/commit/cf47352d71d0e78d39fc634a2d1e659c1030a6ef (SplitWithPredicate())

Only concern is that we lose a bit of context from this PR, so wanted to check with you @TravisEz13 @iSazonov before submitting a new PR off the new branch

iSazonov commented:
Ok, only please add a comment in new PR that it overload the PR.

IISResetMe commented:
Re-submitted as https://github.com/PowerShell/PowerShell/pull/5270, please review. Thanks
